### PR TITLE
Remove duplicated memory_resource_tests

### LINKED ISSUE #1450: Remove duplicated tests that test both a device_memory_resource* and a cuda::mr::async_resource_ref

### PR DESCRIPTION
**Pull request by miscco**
During the initial introduction of `async_resource_ref` we duplicated all tests that take a `device_memory_resource*`.

Now that we have already some experience with running it in production and are moving more of the interfaces to `async resource_ref` remove those duplicated tests.

closes #1450

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/rmm/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: tests/mr/device/mr_ref_tests.cpp
Comment (harrism) [inline tests/mr/device/mr_ref_tests.cpp:36]: **💡 suggestion:** ‏ 

```suggestion
                                           mr_factory{"HostPinned", &make_host_pinned},
                                           mr_factory{"HostPinnedPool", &make_host_pinned_pool},
```

File: tests/mr/device/thrust_allocator_tests.cu
Comment (harrism) [inline tests/mr/device/thrust_allocator_tests.cu:35]: Suggestion. Will need to include `rmm/resource_ref.hpp`

```suggestion
using async_resource_ref = rmm::device_async_resource_ref;
```

**Review by harrism — COMMENTED** [review]
Nice elimination of duplicates. This will save compile and test time!

Suggesting we test the host-pinned MR independently of the host-pinned pool if possible.

File: tests/mr/device/mr_ref_tests.cpp
Comment (miscco) [reply tests/mr/device/mr_ref_tests.cpp:36]: This is actually not possible with the current test setup. 

The tests machinery currently requires a `std::function<std::shared_ptr<rmm::mr::device_memory_resource>>`. However, `pinned_host_memory_resource` is not derived from `device_memory_resource` so that wont compile.

I am wary of completely rewriting the test infra in this PR.

Comment (harrism) [reply tests/mr/device/mr_ref_tests.cpp:36]: Oh I thought when you created device_mr_ref_tests that you had changed that machinery for the ref tests. OK, that can be a follow up.

**Review by harrism — APPROVED** [review]

File: tests/mr/device/mr_ref_tests.cpp
Comment (miscco) [reply tests/mr/device/mr_ref_tests.cpp:36]: The issue is that we need to store the resource somewhere, as `resource_ref` is only a reference and now owning.

That is why I kept most of the base machinery unchanged

Comment (harrism) [reply tests/mr/device/mr_ref_tests.cpp:36]: OK. Seems we need to find a way to test both the (new) MR(s) that just implement the concepts and the (legacy) MRs that just derive from device_memory_resource.

**Comment by miscco** [comment]
Ping for a second review

File: tests/mr/device/mr_ref_tests.cpp
Comment (bdice) [reply tests/mr/device/mr_ref_tests.cpp:36]: @miscco did you have further thoughts here?

> Seems we need to find a way to test both the (new) MR(s) that just implement the concepts and the (legacy) MRs that just derive from device_memory_resource.

**Review by bdice — APPROVED** [review]

**Comment by bdice** [comment]
@miscco I granted a second approval here -- do you need anything else before merging this?

**Comment by miscco** [comment]
I was pondering whether we want to keep the old test, but we are indeed testing `pinned_memory_resource` in the host test suite so there is no need for additional tests AFAIK

**Pull request merged**
